File: lintcode/590.connecting-graph-ii.cpp
class ConnectingGraph2 {
private:
    vector<int> graph;
    vector<int> size;

public:
    /*
    * @param n: An integer
    */ConnectingGraph2(int n) {
        // do intialization if necessary
        graph.resize(n + 1);
        size.resize(n + 1);

        for (int i = 1; i < n + 1; ++i)
        {
            size[i] = 1;
            graph[i] = i;
        }
    }

    int find(int node)
    {
        if (graph[node] == node)
        {
            return node;
        }

        graph[node] = find(graph[node]);
        return graph[node];
    }

    /*
     * @param a: An integer
     * @param b: An integer
     * @return: nothing
     */
    void connect(int a, int b) {
        // write your code here
        int root_a = find(a);
        int root_b = find(b);

        if (root_a != root_b)
        {
            graph[root_a] = root_b;
            size[root_b] += size[root_a]; 
        }
    }

    /*
     * @param a: An integer
     * @return: An integer
     */
    int query(int a) {
        // write your code here
        int root_a = find(a);
        return size[root_a];
    }
};

File: lintcode/590.connecting-graph-ii.py
class ConnectingGraph2:
    """
    @param: n: An integer
    """
    def __init__(self, n):
        # do intialization if necessary
        self.graph = {}
        self.size = {}

        for i in range(1, n + 1):
            self.size[i] = 1
            self.graph[i] = i
    
    def find(self, node):
        if self.graph[node] == node:
            return node

        self.graph[node] = self.find(self.graph[node])
        return self.graph[node]

    """
    @param: a: An integer
    @param: b: An integer
    @return: nothing
    """
    def connect(self, a, b):
        # write your code here
        root_a = self.find(a)
        root_b = self.find(b)
        if root_a != root_b:
            self.graph[root_a] = root_b
            self.size[root_b] += self.size[root_a]


    """ 
    @param: a: An integer
    @return: An integer
    """
    def query(self, a):
        # write your code here
        root_a = self.find(a)
        return self.size[root_a]